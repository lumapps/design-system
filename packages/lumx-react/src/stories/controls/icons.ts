import {
    mdiAbTesting,
    mdiAbjadArabic,
    mdiAccount,
    mdiAccountBox,
    mdiAlert,
    mdiAlertCircle,
    mdiArrowDown,
    mdiArrowUp,
    mdiAtom,
    mdiBee,
    mdiBell,
    mdiBullhornOutline,
    mdiCheck,
    mdiCheckCircle,
    mdiChevronDown,
    mdiChevronLeft,
    mdiChevronRight,
    mdiChevronUp,
    mdiClose,
    mdiCloseCircle,
    mdiDelete,
    mdiDotsHorizontal,
    mdiDragVertical,
    mdiEarth,
    mdiEmail,
    mdiEye,
    mdiFileEdit,
    mdiFlag,
    mdiFolder,
    mdiFolderGoogleDrive,
    mdiFoodApple,
    mdiGoogleCirclesExtended,
    mdiHeart,
    mdiHome,
    mdiImageBroken,
    mdiInformation,
    mdiLink,
    mdiMagnifyMinusOutline,
    mdiMagnifyPlusOutline,
    mdiMenuDown,
    mdiMessageTextOutline,
    mdiMinus,
    mdiOpenInNew,
    mdiPauseCircleOutline,
    mdiPencil,
    mdiPlay,
    mdiPlayCircleOutline,
    mdiPlus,
    mdiRadioboxBlank,
    mdiRadioboxMarked,
    mdiReply,
    mdiSend,
    mdiStar,
    mdiTextBox,
    mdiTextBoxPlus,
    mdiTram,
    mdiTranslate,
    mdiViewList,
} from '@lumx/icons';

// Small selection of mdi icons to
export const iconArgType = {
    control: { type: 'select' },
    options: {
        mdiAbTesting,
        mdiAbjadArabic,
        mdiAccount,
        mdiAccountBox,
        mdiAlert,
        mdiAlertCircle,
        mdiArrowDown,
        mdiArrowUp,
        mdiAtom,
        mdiBee,
        mdiBell,
        mdiBullhornOutline,
        mdiCheck,
        mdiCheckCircle,
        mdiChevronDown,
        mdiChevronLeft,
        mdiChevronRight,
        mdiChevronUp,
        mdiClose,
        mdiCloseCircle,
        mdiDelete,
        mdiDotsHorizontal,
        mdiDragVertical,
        mdiEarth,
        mdiEmail,
        mdiEye,
        mdiFileEdit,
        mdiFlag,
        mdiFolder,
        mdiFolderGoogleDrive,
        mdiFoodApple,
        mdiGoogleCirclesExtended,
        mdiHeart,
        mdiHome,
        mdiImageBroken,
        mdiInformation,
        mdiLink,
        mdiMagnifyMinusOutline,
        mdiMagnifyPlusOutline,
        mdiMenuDown,
        mdiMessageTextOutline,
        mdiMinus,
        mdiOpenInNew,
        mdiPauseCircleOutline,
        mdiPencil,
        mdiPlay,
        mdiPlayCircleOutline,
        mdiPlus,
        mdiRadioboxBlank,
        mdiRadioboxMarked,
        mdiReply,
        mdiSend,
        mdiStar,
        mdiTextBox,
        mdiTextBoxPlus,
        mdiTram,
        mdiTranslate,
        mdiViewList,
    },
};
