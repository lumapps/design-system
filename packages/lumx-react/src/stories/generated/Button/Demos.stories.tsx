/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/button/Demos' };

export { App as EmphasisHigh } from './emphasis-high';
export { App as EmphasisLow } from './emphasis-low';
export { App as EmphasisMedium } from './emphasis-medium';
export { App as Small } from './small';
export { App as Toggle } from './toggle';
