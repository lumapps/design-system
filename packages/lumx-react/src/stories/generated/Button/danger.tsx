../../../../../site-demo/content/product/components/button/react/danger.tsx