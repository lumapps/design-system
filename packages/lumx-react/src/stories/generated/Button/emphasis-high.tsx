../../../../../site-demo/content/product/components/button/react/emphasis-high.tsx