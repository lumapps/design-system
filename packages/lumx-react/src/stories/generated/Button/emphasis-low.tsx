../../../../../site-demo/content/product/components/button/react/emphasis-low.tsx