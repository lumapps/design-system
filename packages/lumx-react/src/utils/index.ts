/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
