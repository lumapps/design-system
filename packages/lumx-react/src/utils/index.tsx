export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
