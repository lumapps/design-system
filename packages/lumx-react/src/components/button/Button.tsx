import React, { forwardRef, ReactNode } from 'react';

import classNames from 'classnames';
import isEmpty from 'lodash/isEmpty';

import { Emphasis, Icon, Size, Theme } from '@lumx/react';
import { Comp, getBasicClass, getRootClassName } from '@lumx/react/utils';
import { BaseButtonProps, ButtonRoot } from './ButtonRoot';

/**
 * Button emphasis definition.
 * @deprecated Use Emphasis instead.
 */
export const ButtonEmphasis = Emphasis;

/**
 * Defines the props of the component.
 */
export interface ButtonProps extends BaseButtonProps {
    /** Button content. */
    children?: ReactNode;
    /** Left icon (SVG path). */
    leftIcon?: string;
    /** Right icon (SVG path). */
    rightIcon?: string;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Button';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<ButtonProps> = {
    emphasis: Emphasis.high,
    size: Size.m,
    theme: Theme.light,
};

/**
 * Button component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Button: Comp<ButtonProps, HTMLButtonElement | HTMLAnchorElement> = forwardRef((props, ref) => {
    const { children, className, emphasis, leftIcon, rightIcon, size, theme, ...forwardedProps } = props;

    const buttonClassName = classNames(
        className,
        getBasicClass({ prefix: CLASSNAME, type: 'hasLeftIcon', value: !isEmpty(leftIcon) }),
        getBasicClass({ prefix: CLASSNAME, type: 'hasRightIcon', value: !isEmpty(rightIcon) }),
    );

    return (
        <ButtonRoot
            ref={ref}
            {...{ emphasis, size, theme, ...forwardedProps }}
            className={buttonClassName}
            variant="button"
        >
            {leftIcon && !isEmpty(leftIcon) && <Icon icon={leftIcon} />}
            {children && <span>{children}</span>}
            {rightIcon && !isEmpty(rightIcon) && <Icon icon={rightIcon} />}
        </ButtonRoot>
    );
});
Button.displayName = COMPONENT_NAME;
Button.className = CLASSNAME;
Button.defaultProps = DEFAULT_PROPS;
