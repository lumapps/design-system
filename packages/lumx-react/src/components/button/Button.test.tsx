import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { mdiCheck, mdiChevronDown, mdiPlus } from '@lumx/icons';
import { commonTestsSuite, Wrapper } from '@lumx/react/testing/utils';
import { getBasicClass } from '@lumx/react/utils';
import { Button, ButtonProps } from './Button';

const DEFAULT_PROPS = Button.defaultProps as any;
const CLASSNAME = Button.className as string;

type SetupProps = Partial<ButtonProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (propsOverride: SetupProps = {}, shallowRendering = true) => {
    const props: any = { ...propsOverride };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    const wrapper: Wrapper = renderer(<Button {...props} />);

    return {
        buttonRoot: wrapper.find('ButtonRoot'),
        icon: wrapper.find('Icon'),
        props,
        wrapper,
    };
};

describe(`<${Button.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        it('should render button with label', () => {
            const label = 'Label';
            const { buttonRoot, icon, wrapper } = setup({
                children: label,
            });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();

            expect(icon).not.toExist();

            expect(buttonRoot.contains(label)).toBe(true);
        });

        it('should render button with label and right icon', () => {
            const label = 'Label';
            const { buttonRoot, icon, wrapper } = setup({
                children: label,
                rightIcon: mdiChevronDown,
            });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();

            expect(icon).toExist();
            expect(icon.length).toEqual(1);

            expect(buttonRoot.contains(label)).toBe(true);
        });

        it('should render button with label and left icon', () => {
            const label = 'Label';
            const { buttonRoot, icon, wrapper } = setup({
                children: label,
                leftIcon: mdiChevronDown,
            });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();

            expect(icon).toExist();
            expect(icon.length).toEqual(1);

            expect(buttonRoot.contains(label)).toBe(true);
        });

        it('should render button with label and icons', () => {
            const label = 'Label';
            const { buttonRoot, icon, wrapper } = setup({
                children: label,
                leftIcon: mdiCheck,
                rightIcon: mdiChevronDown,
            });
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toExist();

            expect(icon).toExist();
            expect(icon.length).toEqual(2);

            expect(buttonRoot.contains(label)).toBe(true);
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        it('should use default props', () => {
            const { buttonRoot } = setup();

            const actualProps = buttonRoot.props() as Partial<ButtonProps>;
            expect(actualProps.variant).toEqual('button');
            for (const [propName, propValue] of Object.entries(DEFAULT_PROPS)) {
                expect(actualProps[propName]).toEqual(propValue);
            }
        });

        it('should forward any CSS class', () => {
            const props: Partial<ButtonProps> = {
                className: 'component component--is-tested',
            };
            const { wrapper, buttonRoot } = setup(props);
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toHaveClassName(props.className);
        });

        it('should use the given props for class names', () => {
            const props: Partial<ButtonProps> = {
                leftIcon: mdiChevronDown,
                rightIcon: mdiPlus,
            };
            const { wrapper, buttonRoot } = setup(props);
            expect(wrapper).toMatchSnapshot();

            expect(buttonRoot).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'hasLeftIcon', value: true }));
            expect(buttonRoot).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'hasRightIcon', value: true }));
        });
    });

    // 3. Test events.
    describe('Events', () => {
        // Nothing to do here.
    });

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        // Nothing to do here.
    });

    // 5. Test state.
    describe('State', () => {
        // Nothing to do here.
    });

    // Common tests suite.
    commonTestsSuite(setup, { prop: 'buttonRoot' }, { className: CLASSNAME });
});
