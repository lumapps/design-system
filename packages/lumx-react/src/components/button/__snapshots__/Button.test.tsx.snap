// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> Props should forward any CSS class 1`] = `
<ButtonRoot
  className="component component--is-tested"
  emphasis="high"
  size="m"
  theme="light"
  variant="button"
/>
`;

exports[`<Button> Props should use the given props for class names 1`] = `
<ButtonRoot
  className="lumx-button--has-left-icon lumx-button--has-right-icon"
  emphasis="high"
  size="m"
  theme="light"
  variant="button"
>
  <Icon
    color="dark"
    icon="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
    size="m"
  />
  <Icon
    color="dark"
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    size="m"
  />
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render button with label 1`] = `
<ButtonRoot
  className=""
  emphasis="high"
  size="m"
  theme="light"
  variant="button"
>
  <span>
    Label
  </span>
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render button with label and icons 1`] = `
<ButtonRoot
  className="lumx-button--has-left-icon lumx-button--has-right-icon"
  emphasis="high"
  size="m"
  theme="light"
  variant="button"
>
  <Icon
    color="dark"
    icon="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
    size="m"
  />
  <span>
    Label
  </span>
  <Icon
    color="dark"
    icon="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
    size="m"
  />
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render button with label and left icon 1`] = `
<ButtonRoot
  className="lumx-button--has-left-icon"
  emphasis="high"
  size="m"
  theme="light"
  variant="button"
>
  <Icon
    color="dark"
    icon="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
    size="m"
  />
  <span>
    Label
  </span>
</ButtonRoot>
`;

exports[`<Button> Snapshots and structure should render button with label and right icon 1`] = `
<ButtonRoot
  className="lumx-button--has-right-icon"
  emphasis="high"
  size="m"
  theme="light"
  variant="button"
>
  <span>
    Label
  </span>
  <Icon
    color="dark"
    icon="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
    size="m"
  />
</ButtonRoot>
`;
